cpp/ql/src/semmle/code/cpp/ir/dataflow/DefaultTaintTracking.qll

```diff
@@ -37,6 +37,61 @@ private class DefaultTaintTrackingCfg extends DataFlow::Configuration {
   }
 }
 
+private class ToGlobalVarTaintTrackingCfg extends DataFlow::Configuration {
+  ToGlobalVarTaintTrackingCfg() { this = "GlobalVarTaintTrackingCfg" }
+
+  override predicate isSource(DataFlow::Node source) { isUserInput(source.asExpr(), _) }
+
+  override predicate isSink(DataFlow::Node sink) {
+    exists(GlobalOrNamespaceVariable gv |
+      accessesVariable(sink.asInstruction(), gv) and
+      sink.asInstruction() instanceof StoreInstruction
+    )
+  }
+
+  override predicate isAdditionalFlowStep(DataFlow::Node n1, DataFlow::Node n2) {
+    instructionTaintStep(n1.asInstruction(), n2.asInstruction())
+    or
+    exists(StoreInstruction i1, LoadInstruction i2, GlobalOrNamespaceVariable gv |
+      accessesVariable(i1, gv) and
+      accessesVariable(i2, gv) and
```

Comment:
This is a dangerous cross product, but I don't see any easy way to get rid of it completely. Unless the optimiser does something great here, I think we can reduce the cross-product risk a lot by in-lining `accessesVariable` here -- we're only using one of its two cases for each of i1 and i2.

To be more precise, I'll suggest (for here and elsewhere) to factor `accessesVariable` into a `readsVariable` and `writesVariable`.